### PR TITLE
Add SPI Flash based bootstrap support

### PR DESCRIPTION
**Pull request by alphan**
This PR
- removes `cw_spiflash.py` vendored from lowRISC/opentitan, and
- adds SPI Flash based bootstrap support.

We should probably wait until lowRISC/opentitan#12875 is merged and a bitstream with the new bootstrap protocol is added to this repo before merging this PR.

Also, I couldn't try this change thoroughly since I don't have a scope with me right now but I was able to bootstrap the target using `./capture.py init` and commenting out the line related to scope in `device.py`. I'd appreciate if someone else could also give this a try before merging.

### CODE REVIEW

File: cw/cw305/util/spiflash.py
Comment (moidx) [inline cw/cw305/util/spiflash.py:117]: nit: It took me a bit to figure out what this does. I think it would be good to add more details to the documentation to describe that at the end of bootstrap there is an additional reset.

File: cw/cw305/util/spiflash.py
Comment (moidx) [inline cw/cw305/util/spiflash.py:126]: nit: rm deadcode

File: cw/cw305/util/spiflash.py
Comment (moidx) [inline cw/cw305/util/spiflash.py:139]: s/chip_erase/chip_erase_and_write_enable/

**Review by moidx — APPROVED** [review]

**Comment by vogelpi** [comment]
Thanks for the PR @alphan! Please wait with merging this. I am a little behind here and first need to generate new bitstreams.

**Comment by vogelpi** [comment]
Hi @alphan ,
I have this a try with updated binaries. I've directly pushed them into this PR. The CW310 works fine but the CW305 is currently broken. I get the following output:
```
./capture.py --cfg-file capture_aes_cw305.yaml init
WARNING:ChipWhisperer Target:No FPGA Bitstream file specified.
Connecting and loading FPGA... Done!
Initializing PLL1
WARNING:ChipWhisperer Scope:
                ADC frequency exceeds specification (200 MHz). 
                This may or may not work, depending on temperature, voltage, and luck.
                It may not work reliably.
                You can run scope.adc_test() to check whether ADC data is sampled properly by the FPGA,
                but this doesn't fully verify that the ADC is working properly.
                
Traceback (most recent call last):
  File "./capture.py", line 652, in <module>
    app()
  File "/home/pirmin/ot/ot-sca/.venv/lib/python3.8/site-packages/typer/main.py", line 214, in __call__
    return get_command(self)(*args, **kwargs)
  File "/home/pirmin/ot/ot-sca/.venv/lib/python3.8/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/home/pirmin/ot/ot-sca/.venv/lib/python3.8/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/home/pirmin/ot/ot-sca/.venv/lib/python3.8/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/pirmin/ot/ot-sca/.venv/lib/python3.8/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/pirmin/ot/ot-sca/.venv/lib/python3.8/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/home/pirmin/ot/ot-sca/.venv/lib/python3.8/site-packages/typer/main.py", line 500, in wrapper
    return callback(**use_params)  # type: ignore
  File "./capture.py", line 121, in init
    initialize_capture(ctx.obj.cfg["device"], ctx.obj.cfg["capture"])
  File "./capture.py", line 64, in initialize_capture
    ot = device.OpenTitan(device_cfg["fpga_bitstream"],
  File "/home/pirmin/ot/ot-sca/cw/cw305/util/device.py", line 60, in __init__
    self.target = self.initialize_target(programmer, firmware, baudrate, output_len)
  File "/home/pirmin/ot/ot-sca/cw/cw305/util/device.py", line 148, in initialize_target
    programmer.bootstrap(firmware)
  File "/home/pirmin/ot/ot-sca/cw/cw305/util/spiflash.py", line 174, in bootstrap
    self.chip_erase()
  File "/home/pirmin/ot/ot-sca/cw/cw305/util/spiflash.py", line 146, in chip_erase
    self.busy_poll()
  File "/home/pirmin/ot/ot-sca/cw/cw305/util/spiflash.py", line 165, in busy_poll
    raise RuntimeError("Reached maximum iteration count")
RuntimeError: Reached maximum iteration count
```
I don't know what exactly the issue is. More debugging is needed. It could be that the bitstreams we currently produce are non-functional. It's hard to say and difficult to debug - especially as so many things changed recently.

Before merging we should squash the three commits into to get an atomic commit.

**Comment by vogelpi** [comment]
Connecting the UART to the debug output gives me this output:
```
I00000 test_rom.c:81] Version:    earlgrey_silver_release_v5-5864-g64f9a9a90
Build Date: 2022-06-10, 15:10:34

I00001 test_rom.c:87] TestROM:532d040c
I00002 test_rom.c:118] Test ROM complete, jumping to flash!
```
So indeed, the FPGA bitstream is functional but the bootstrapping doesn't  seem to work. Do you have any ideas @alphan ?

**Comment by alphan** [comment]
Thanks for giving this a try @vogelpi! I remember trying this with cw305 but did so only locally before merging the PR. I'll try with cw305 but might not be able to get back to this until mid next week. Just to make sure, is this blocking you right now?

**Comment by vogelpi** [comment]
Thanks for the quick response @alphan. Yes, this is blocking. We have done quite some work on the setup that requires new aes/sha_serial firmware binaries. But since the current firmware is no longer compatible with the old bitstreams we need to update everything to rerun experiments. If you could have a look this would be great. I am happy to assist with testing if you no longer have access to the board/UART debugger etc.

**Comment by alphan** [comment]
> Thanks for the quick response @alphan. Yes, this is blocking. We have done quite some work on the setup that requires new aes/sha_serial firmware binaries. But since the current firmware is no longer compatible with the old bitstreams we need to update everything to rerun experiments. If you could have a look this would be great. I am happy to assist with testing if you no longer have access to the board/UART debugger etc.

Just saw your message, looking at this now.

**Comment by alphan** [comment]
> `WARNING:ChipWhisperer Target:No FPGA Bitstream file specified.`

I wonder where this comes from...

**Comment by alphan** [comment]
> Connecting the UART to the debug output gives me this output:
> 
> ```
> I00000 test_rom.c:81] Version:    earlgrey_silver_release_v5-5864-g64f9a9a90
> Build Date: 2022-06-10, 15:10:34
> 
> I00001 test_rom.c:87] TestROM:532d040c
> I00002 test_rom.c:118] Test ROM complete, jumping to flash!
> ```
> 
> So indeed, the FPGA bitstream is functional but the bootstrapping doesn't seem to work. Do you have any ideas @alphan ?

This was due to a `flash_ctrl.hjson` that was not updated by `prepare_sw.py`. I'll create a PR in lowRISC/opentitan to fix this. For reference, I now get the following output when I run `./capture.py --cfg-file capture_aes_cw305.yaml init`:

```
I00000 test_rom.c:81] Version:    earlgrey_silver_release_v5-5890-g49010d318
Build Date: 2022-06-11, 15:08:48

I00001 test_rom.c:87] TestROM:5b2cf8cb
I00002 test_rom.c:118] Test ROM complete, jumping to flash!
I00000 aes_serial.c:332] Running AES serial
I00001 aes_serial.c:334] Initializing simple serial interface to capture board.
I00002 aes_serial.c:342] Initializing AES unit.
I00003 aes_serial.c:345] Starting simple serial packet handling.
```

**Comment by alphan** [comment]
lowRISC/opentitan#13173 should unblock this. Could you please update the binaries using that PR and try again?

Also:

> Before merging we should squash the three commits into to get an atomic commit.

I'd prefer keeping the 3 commits separate if you don't mind (since they are all self-contained).

**Comment by alphan** [comment]
The last push addresses @moidx's comments and removes @vogelpi's commit that updates binaries. @vogelpi can you push another commit that updates the binaries (based on lowRISC/opentitan#13173) and let me know if everything works as expected? Thanks!

**Comment by vogelpi** [comment]
Thanks @alphan for fixing this. We can also keep the commits seperate. I will now generate new bitstreams and again push to this PR.

**Comment by alphan** [comment]
Thanks @vogelpi! This should be ready to merge as long as there are no issues. PLMK if you have any problems.

**Review by vogelpi — APPROVED** [review]
Thanks again for your help on this @alphan! I've now updated the bitstreams and binaries and pushed them into this PR. LGTM!

**Pull request merged**
